// messaging-api-line/lib/Line.js
'use strict';Object.defineProperty(exports, "__esModule", { value: true });










function createText(text) {
  return {
    type: 'text',
    text };

}

function createImage(contentUrl, previewUrl) {
  return {
    type: 'image',
    originalContentUrl: contentUrl,
    previewImageUrl: previewUrl || contentUrl };

}

function createVideo(contentUrl, previewUrl) {
  return {
    type: 'video',
    originalContentUrl: contentUrl,
    previewImageUrl: previewUrl };

}

function createAudio(contentUrl, duration) {
  return {
    type: 'audio',
    originalContentUrl: contentUrl,
    duration };

}

function createLocation({ title, address, latitude, longitude }) {
  return {
    type: 'location',
    title,
    address,
    latitude,
    longitude };

}

function createSticker(packageId, stickerId) {
  return {
    type: 'sticker',
    packageId,
    stickerId };

}

function createImagemap(
altText,
{
  baseUrl,
  baseSize,
  baseHeight,
  baseWidth,
  actions })










{
  return {
    type: 'imagemap',
    baseUrl,
    altText,
    baseSize: baseSize || {
      height: baseHeight,
      width: baseWidth },

    actions };

}

function createTemplate(altText, template) {
  return {
    type: 'template',
    altText,
    template };

}

function createButtonTemplate(
altText,
{
  thumbnailImageUrl,
  imageAspectRatio,
  imageSize,
  imageBackgroundColor,
  title,
  text,
  actions })









{
  return createTemplate(altText, {
    type: 'buttons',
    thumbnailImageUrl,
    imageAspectRatio,
    imageSize,
    imageBackgroundColor,
    title,
    text,
    actions });

}

function createConfirmTemplate(
altText,
{
  text,
  actions })




{
  return createTemplate(altText, {
    type: 'confirm',
    text,
    actions });

}

function createCarouselTemplate(
altText,
columns,
{
  imageAspectRatio,
  imageSize })




{
  return createTemplate(altText, {
    type: 'carousel',
    columns,
    imageAspectRatio,
    imageSize });

}

function createImageCarouselTemplate(
altText,
columns)
{
  return createTemplate(altText, {
    type: 'image_carousel',
    columns });

}

const Line = {
  createText,
  createImage,
  createVideo,
  createAudio,
  createLocation,
  createSticker,
  createImagemap,
  createTemplate,
  createButtonTemplate,
  createConfirmTemplate,
  createCarouselTemplate,
  createImageCarouselTemplate };exports.default =


Line;
